### ISSUE TITLE
attached is being called every time a bound variable is changed in latest version

### ISSUE THREAD

**Issue by mbroadst** [aurelia/templating#128]
In previous versions of aurelia, attached would only be called the first time my custom element was attached to the DOM. When I had my elements generated from a list of objects, each object would create a new element and a single attached event would occur for each of them (so I would set up my DOM related work here). Each time that backing list was updated (not appended to, but items being replaced) I would not have an additional "attached" call, presumably because the element was not in fact being recreated but instead modified. 

This no longer seems to be the case with the latest release. Instead, each time I change the list I get an attached no matter what. The plunkers aren't updated, so I can't provide one of those, but if you take a look at the following commit to my skeleton-navigation project you can see an example of how to reproduce this error: https://github.com/mbroadst/skeleton-navigation/commit/e149e7aae9ae9c683abe622f446faa387a6521b6

TL;DR - if you look at the commit I posted (or just check out the project), you'll note that "attached called!" will be emitted every single time some modification occurs to the list of "things". Compared with this plunker: http://plnkr.co/edit/moBBoMPlKP2K0EOE8YGW, you'll note that this was not the same behavior in previous versions of aurelia.

**Comment by EisenbergEffect** [comment]
@martingust Can you investigate? I believe this relates to the fact that the correct view lifecycle is now preserved even when views are re-used.

**Comment by mbroadst** [comment]
@EisenbergEffect the posted examples are a bit contrived, in my actual use case I have an aurelia element which wraps d3 calls, creating a widget for a graph. Ideally, I would like to create the graph when the element is bound and attached to the DOM the first time, and on subsequent updates simply change parts of the SVG. Is there a better lifecycle event to tap into? (also is this the correct repo for this issue?)

**Comment by mbroadst** [comment]
@EisenbergEffect in re-reading your comment I believe the opposite is the case, in that the lifecycle is not preserved whereas it used to be

**Comment by stoffeastrom** [comment]
@mbroadst In that plunker you are recreating the bound values on every interval. What happens if you keep the object references and only update the properties? As @EisenbergEffect said the *view lifecycle* is now fully preserved which means that since it's new object references the view will be treated as it's recreated but keeping the DOM element.

**Comment by mbroadst** [comment]
@stoffeastrom ah okay I see my mistake in interpretation there. Yeah, I can definitely merge incoming changes (or my temporary approach is just to use an 'initialized' flag in the widget to determine if its already happened), but isn't that sort of one of the points of using such a framework?

Aurelia already has a powerful engine for dirty checking that I should be able to leverage for this case. Furthermore it seems to follow that if "attached" indicates that the "view that contains the extension is attached to the DOM," then changing the bound value (whether its a brand new value, or a change in a referenced value) should not in fact trigger a reattach since the DOM should not be changing? Perhaps I still don't fully understand what lifecycle refers to in this case.

**Comment by stoffeastrom** [comment]
@mbroadst I agree regarding the reattach. Had a long discussion https://github.com/aurelia/templating-resources/issues/45 about the *view lifecycle* and there might still be some kind of bugs around it.

**Comment by martingust** [comment]
@mbroadst I tested your repo and that is the expected behavior after the latest release where the view lifecycle is fully preserved. It's also same behavior as before we started to re-use views in the repeater. 

However I do agree with you all that we should look into the view lifecycle behavior in the repeat. We have some related work to do in there when implementing view caching and should definitely look into this at the same time.

**Comment by mbroadst** [comment]
@martingust I'm getting some weird behavior now where my refs within the repeat elements are being lost. I haven't dug too deeply into it, but it renders my graphs (I have a repeat.for of d3 graphs) useless when the repeat.for element is reattached. Is this also expected behavior? Have you experienced this?

**Comment by martingust** [comment]
@mbroadst That is related and an issue. The call to unbind nulls out the reference to prevent memory leaks, but is obviously not correct behavior for your scenario. We are working on the `view caching` implementation which should be in place this week, after that we are going to revisit this issue and dig into your use case to see what is needed to be done. Sounds ok?

**Comment by mbroadst** [comment]
@martingust sure that sounds great. I've been able to get around the issue in my app by implementing an "initialized" boolean to prevent further attaches from initializing the view, and by injecting Element and no longer using refs.

To be honest, I tried to modify the example I posted above in a plunker to show that refs were being lost, but I just couldn't create a small test case exposing this issue. For now I guess I'll have to take your word for it and look out for changes in the next update :smile:

**Comment by martingust** [comment]
@mbroadst Great. I can't promise it will go out in the next update though, after view caching is in place we will start the work inside the repeat, can only promise we are going to get it done as soon as we can :)

**Comment by AureliaEffect** [comment]
@mbroadst I'm going to be working on a view cache mechanism in Aurelia's core this week. I'd love to hear a bit more about these issues. Is it primarily related to "re-attaching". If so, we might be able to have the new view cache mechanism tell you in the `attach` callback that the view is being re-used, for example. Would something like that be beneficial?

**Comment by stoffeastrom** [comment]
haha *AureliaEffect*, now we`re talking!!

**Comment by mbroadst** [comment]
@EisenbergEffect is that you? heh

Yes that feature would certainly solve my two use cases. As I mentioned I tried (in vain) for about 20min to get you guys a test case on the plunkers exhibiting the behavior of losing my refs, but couldn't get it to happen there. Seems like @martingust has a handle on that though. 

I also think it's important to specify in the documentation that the attached lifecycle event is very specifically about the _view_ being attached to the DOM (which appears to be different than the contents of the view being _added_ to the DOM). It's rather confusing when I assume that attached means that the contents of my view have been added and attached to the DOM, that somehow the view would be reattached because its backing data has changed (even though nothing on the DOM has been added or removed).

**Comment by martingust** [comment]
@mbroadst @AureliaEffect Here is a gh-pages that shows the issue http://martingust.github.io/ref-issue/. 

Open console and hit the button. Two changes in the repeat will get logged. One that triggers a view re-use where the ref element is null and one with a new view where the element is referenced.

**Comment by mbroadst** [comment]
hey guys, I found another related bug, please let me know if you'd like a new issue opened for this. 

[this example](http://plnkr.co/edit/pbSbFRl2dEiUTKsrixPO?p=preview) is broken in firefox and safari on my mac here. The reason it's related is because I'm changing individual elements of the list backing the repeat.for, and they are not updating on those two browsers. It's worth noting that if you replaced the entire backing list wholesale this error does not occur (my current workaround).

**Comment by mbroadst** [comment]
This is actually a much bigger issue for me than I thought. While replacing the entire backing array wholesale works and looks fine in the given plunker, when I use the same approach on my page (which draws a list of d3 graphs based on this array) I get a huge flicker as it lays out all the DOM elements again.. Any thoughts on how to work around this in the interim?

**Comment by EisenbergEffect** [comment]
@mbroadst You can't update arrays by index without native Object.observe. It is the only limitation of our polyfill. It's not possible to fix it. That's why you see the problem on Safari and Firefox, but not Chrome. Instead, you can use the splice method. That's the recommended approach.

**Comment by mbroadst** [comment]
@EisenbergEffect okay thanks. It's worth noting for future people looking over this that splice only seems to work on an individual item (not splicing the entire array)
